### PR TITLE
Shortens blocks' last_key in the SSTable block index.

### PR DESCRIPTION
**Pull request by fulmicoton**
Right now we store last key in the blocks of the SSTable index.
This PR replaces the last key by a shorter string that is greater or
equal and still lesser than the next key.
This property is sufficiently to ensure the block index
works properly.

### CODE REVIEW

File: src/termdict/sstable_termdict/sstable/sstable_index.rs [hunk 1]
```diff
@@ -41,10 +44,40 @@ pub struct SSTableIndexBuilder {
...
+fn find_shorter_str_in_between(left: &mut Vec<u8>, right: &[u8]) {
+    assert!(&left[..] < right);
+    let common_len = common_prefix_len(&left, right);
+    if left.len() == common_len {
+        return;
+    }
+    // It is possible to do one character shorter in some case,
+    // but it is not worth the extra complexity
+    for pos in (common_len + 1)..left.len() {
+        if left[pos] != 255u8 {
```
Comment (PSeitz) [inline src/termdict/sstable_termdict/sstable/sstable_index.rs]: what's the meaning of 255 here?

the comment says string, but we operate on bytes, what about utf-8?

Comment (fulmicoton) [reply src/termdict/sstable_termdict/sstable/sstable_index.rs]: 255 is the max value for a u8.

It is not especially utf8. All of these are byte strings.
There are already u64 byte representation in there.

We do not expose the last term in block so it should be safe.

File: src/termdict/sstable_termdict/sstable/sstable_index.rs [hunk 2]
```diff
@@ -41,10 +44,39 @@ pub struct SSTableIndexBuilder {
...
```
Comment (PSeitz) [inline src/termdict/sstable_termdict/sstable/sstable_index.rs]:
```suggestion
        if left[pos] != u8::MAX {
```

File: src/termdict/sstable_termdict/sstable/mod.rs [hunk 3]
```diff
@@ -145,6 +145,12 @@ where
     }
 
     pub fn write_key(&mut self, key: &[u8]) {
+        // If this is the first key in the block, we use it to
+        // shorten the last term in block in the last block.
```
Comment (PSeitz) [inline src/termdict/sstable_termdict/sstable/mod.rs]:
```suggestion
        // shorten the last term in the last block.
```

File: src/termdict/sstable_termdict/sstable/sstable_index.rs [hunk 4]
```diff
@@ -32,7 +33,9 @@ pub struct BlockAddr {
 
 #[derive(Debug, Serialize, Deserialize)]
 struct BlockMeta {
-    pub last_key: Vec<u8>,
+    /// Any byte string that is lexicographically greater than the last key in the block,
```
Comment (PSeitz) [inline src/termdict/sstable_termdict/sstable/sstable_index.rs:36]: greater than or equal to?

**Review by PSeitz — APPROVED** [review]

File: src/termdict/sstable_termdict/sstable/sstable_index.rs (hunk 4 above)
Comment (fulmicoton) [reply src/termdict/sstable_termdict/sstable/sstable_index.rs:36]: yes absolutely. fixing.

**Pull request merged**